contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,381 @@
+  Kubernetes version that uses CSI drivers for in-tree volume plugins to an old
+  Kubernetes version that uses old-fashioned volume plugins without CSI.
+* Design framework for migration that allows for easy interface extension by
+  in-tree plugin authors to “migrate” their plugins.
+    * Migration must be modular so that each plugin can have migration turned on
+      and off separately
+
+## Non-Goals
+* Design a mechanism for deploying  CSI drivers on all systems so that users can
+  use the current storage system the same way they do today without having to do
+  extra set up.
+* Implementing CSI Drivers for existing plugins
+* Define set of volume plugins that should be migrated to CSI
+
+## Implementation Schedule
+
+Alpha [1.13]
+* Off by default
+* Proof of concept migration of at least 2 storage plugins [AWS, GCE]
+* Framework for plugin migration built for Dynamic provisioning, pre-provisioned
+  volumes, and in-tree volumes
+
+Beta [Target 1.15]
+* On by default
+* Migrate all of the cloud provider plugins*
+
+GA [TBD]
+* Feature on by default, per-plugin toggle on for relevant cloud provider by
+  default
+* CSI Drivers for migrated plugins available on related cloud provider cluster
+  by default
+
+## Feature Gating
+We will have an alpha feature gate for the whole feature that can turn the CSI
+migration on or off, when off all code paths should revert/stay with the in-tree
+plugins. We will also have individual flags for each driver so that admins can
+toggle them on or off.
+
+The feature gate can exist at the interception points in the OperationGenerator
+for Attach and Mount, as well as in the PV Controller for Provisioning.
+
+We will also have one feature flag for each driver’s migration so that each
+driver migration can be turned on and off individually. 
+
+The new feature gates for alpha are:
+```
+// Enables the in-tree storage to CSI Plugin migration feature.
+CSIMigration utilfeature.Feature = "CSIMigration"
+
+// Enables the GCE PD in-tree driver to GCE CSI Driver migration feature.
+CSIMigrationGCE utilfeature.Feature = "CSIMigrationGCE"
+
+// Enables the AWS in-tree driver to AWS CSI Driver migration feature.
+CSIMigrationAWS utilfeature.Feature = "CSIMigrationAWS"
+```
+
+## Translation Layer
+The main mechanism we will use to migrate plugins is redirecting in-tree
+operation calls to the CSI Driver instead of the in-tree driver, the external
+components will pick up these in-tree PV's and use a translation library to
+translate to CSI Source.
+
+Pros:
+* Keeps old API objects as they are
+* Facilitates gradual roll-over to CSI
+
+Cons:
+* Somewhat complicated and error prone.
+* Bespoke translation logic for each in-tree plugin
+
+### Dynamically Provisioned Volumes
+
+#### Kubernetes Changes
+Dynamically Provisioned volumes will continue to be provisioned with the in-tree
+`PersistentVolumeSource`. We will create an external provisioner to pick up the
```

Comment:
csi-external provisioner can pick up the volume, and have reverse translation logic to also create in-tree PV's